workflows/cloudify_system_workflows/snapshots/snapshot_restore.py

```diff
@@ -284,16 +328,23 @@ def _restore_plugins(self, existing_plugins):
         """
         ctx.logger.info('Restoring plugins')
         plugins_to_install = self._get_plugins_to_install(existing_plugins)
-        original_tenant_name = self._tenant_name
-        for plugin in plugins_to_install:
-            self._tenant_name = plugin['tenant_name']
-            with self._update_tenant_in_ctx():
-                plugins.install(ctx=ctx, plugin={
-                    'name': plugin['package_name'],
-                    'package_name': plugin['package_name'],
-                    'package_version': plugin['package_version']
-                })
-        self._tenant_name = original_tenant_name
+        for tenant, plugins in plugins_to_install.items():
+            client = get_rest_client(tenant=tenant)
+            plugins_tmp = tempfile.mkdtemp()
+            for plugin in plugins:
+                wagon_name = os.path.basename(plugin['path'])
+                ctx.logger.info(
+                    'Installing plugin {plugin} for {tenant}'.format(
+                        plugin=wagon_name,
+                        tenant=tenant,
+                    )
+                )
+                temp_plugin = os.path.join(plugins_tmp, wagon_name)
+                shutil.copyfile(plugin['path'], temp_plugin)
+                client.plugins.delete(plugin['id'], force=True)
+                client.plugins.upload(temp_plugin)
+                os.remove(temp_plugin)
```

Comment:
Will do- I wasn't so concerned with stuff like that in this because a failed snapshot restore basically results in an inconsistent manager so it almost certainly needs burning down and starting again if anything about the restore fails.